File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RefreshLayout.java
package com.garpr.android.views;


import android.content.Context;
import android.content.res.TypedArray;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AbsListView;
import android.widget.ScrollView;

import com.garpr.android.R;


public class RefreshLayout extends SwipeRefreshLayout {


    private static final int DEFAULT_SPINNER_COLORS = R.array.spinner_colors;

    private int mScrollingChildId;
    private View mScrollingChild;




    public RefreshLayout(final Context context, final AttributeSet attrs) {
        super(context, attrs);
        parseAttributes(attrs);
    }


    @Override
    public boolean canChildScrollUp() {
        final boolean canChildScrollUp;

        if (mScrollingChild == null) {
            canChildScrollUp = super.canChildScrollUp();
        } else {
            // -1 means to check scrolling up (1 would check scrolling down)
            canChildScrollUp = ViewCompat.canScrollVertically(mScrollingChild, -1);
        }

        return canChildScrollUp;
    }


    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        if (mScrollingChildId != 0) {
            final View scrollingChild = findViewById(mScrollingChildId);

            if (scrollingChild == null) {
                throw new IllegalStateException("unable to find scrolling child with view ID of "
                        + mScrollingChild);
            }

            setScrollingChild(scrollingChild);
        }
    }


    private void parseAttributes(final AttributeSet attrs) {
        final TypedArray ta = getContext().obtainStyledAttributes(attrs,
                R.styleable.refresh_layout, 0, 0);

        if (ta.hasValue(R.styleable.refresh_layout_scrolling_child)) {
            mScrollingChildId = ta.getResourceId(R.styleable.refresh_layout_scrolling_child, 0);
        }

        final int spinnerColorsResId = ta.getResourceId(
                R.styleable.refresh_layout_spinner_colors, DEFAULT_SPINNER_COLORS);
        final int spinnerColors[] = getResources().getIntArray(spinnerColorsResId);
        setColorSchemeColors(spinnerColors);

        if (ta.hasValue(R.styleable.refresh_layout_spinner_size)) {
            setSize(ta.getInt(R.styleable.refresh_layout_spinner_size, DEFAULT));
        }

        ta.recycle();
    }


    @Override
    public void setRefreshing(final boolean refreshing) {
        post(new Runnable() {
            @Override
            public void run() {
                RefreshLayout.super.setRefreshing(refreshing);
            }
        });
    }


    public void setScrollingChild(final View scrollingChild) throws IllegalArgumentException {
        if (scrollingChild instanceof AbsListView || scrollingChild instanceof RecyclerView
                || scrollingChild instanceof ScrollView) {
            mScrollingChild = scrollingChild;
        } else {
            throw new IllegalArgumentException("scrollingChild (" + scrollingChild +
                    ") must be an AbsListView, RecyclerView, or ScrollView");
        }
    }


}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RefreshLayout.kt
package com.garpr.android.views

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.annotation.IdRes
import androidx.core.view.ViewCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.garpr.android.R
import com.garpr.android.misc.Heartbeat
import com.garpr.android.misc.ListLayout

/**
 * A child class of the official Android [SwipeRefreshLayout] that helps us work around some of
 * its shortcomings. We should use this view instead of SwipeRefreshLayout in every case.
 */
open class RefreshLayout @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : SwipeRefreshLayout(context, attrs), Heartbeat, ListLayout {

    @IdRes
    private var scrollingChildId: Int = View.NO_ID


    init {
        parseAttributes(attrs)
    }

    /*
     * http://stackoverflow.com/q/25270171/823952
     */
    override fun canChildScrollUp(): Boolean {
        return scrollingChild?.canScrollVertically(-1) ?: super.canChildScrollUp()
    }

    override val isAlive: Boolean
        get() = ViewCompat.isAttachedToWindow(this)

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (scrollingChildId != View.NO_ID) {
            scrollingChild = findViewById(scrollingChildId) ?: throw NullPointerException(
                    "unable to find scrolling child (${resources.getResourceName(scrollingChildId)})")
        }

        setProgressBackgroundColorSchemeResource(R.color.card_background)
    }

    private fun parseAttributes(attrs: AttributeSet?) {
        if (isInEditMode) {
            return
        }

        val ta = context.obtainStyledAttributes(attrs, R.styleable.RefreshLayout)
        val spinnerColorsResId = ta.getResourceId(R.styleable.RefreshLayout_spinnerColors,
                R.array.spinner_colors)
        setColorSchemeColors(*resources.getIntArray(spinnerColorsResId))

        if (ta.hasValue(R.styleable.RefreshLayout_scrollingChildId)) {
            scrollingChildId = ta.getResourceId(R.styleable.RefreshLayout_scrollingChildId, View.NO_ID)
        }

        ta.recycle()
    }

    override val recyclerView: RecyclerView?
        get() = scrollingChild as? RecyclerView

    var scrollingChild: View? = null

}
